build/release: create new release package for reproducible builds (#3545)
lightningnetwork/lnd: build/release/README.md

# lnd's Reproducible Build System

> This package contains the container that the `lnd` project uses in order to

Reviewer: Can remove the container reference.

build binaries for each new release. As of Go 1.13, with some new build flags,
binaries are now reproducible, allowing developers to build the binary on
distinct machines, and end up with a byte-for-byte identical binary.  However,
this wasn't _fully_ solved in Go 1.13, as the build system still includes the

> directory the binary is built into the binary itself. As a result, our scripts

Author: Well we'll only do our "official" builds w/ the latest one at the time likely, so after Go 1.13.2 re can remove the buildid stuff, but they'll still need to use that latest version.

utilize a work around needed until Go 1.13.2.  

## Building a New Release

### MacOS/Linux/Windows (WSL)

No prior set up is needed on Linux or MacOS is required in order to build the
release binaries.  However, on Windows, the only way to build the binaries atm
is using the Windows Subsystem Linux. You can build the release binaries from

> your `lnd` directory, no matter where it's located.  One can download `lnd` to

Reviewer: Nit: inconsistent use of "one" and "you" 😛

Author: Must've been written at distinct times my two of my distinct selves.

```
git clone https://github.com/lightningnetwork/lnd.git
```

Afterwards, the release manager/verifier simply needs to run:
`./build/release/release.sh <TAG>` (from the top-level `lnd` directory) , where
`<TAG>` is the name of the next release/tag.

## Verifying a Release

With Go 1.13, it's now possible for third parties to verify a release binary.
Before this release, one had to trust that release manager to build the proper
binary. With this new system, third parties can now _independently_ run the
release process, and verify that all the hashes in the final `manifest.txt`

> match up.

Reviewer 2: should we also note that verification must be done using the same go version?

Author: Was thinking to leave it to the release notes to dictate which version is used so we don't need to update this README each time a new version is released.

Reviewer 2: gotcha, yeah i was thinking here we'd just say something like "when verifying, you must use the same version of go as the releaser. check the release notes for the exact version of go to use for a particular set of binaries."

# lnd's Reproducible Build System

This package contains the container that the `lnd` project uses in order to
build binaries for each new release. As of Go 1.13, with some new build flags,
binaries are now reproducible, allowing developers to build the binary on
distinct machines, and end up with a byte-for-byte identical binary.  However,
this wasn't _fully_ solved in Go 1.13, as the build system still includes the
directory the binary is built into the binary itself. As a result, our scripts
utilize a work around needed until Go 1.13.2.  

## Building a New Release

> ### MacOS/Windows

Reviewer: Windows (Git Bash) is the only one that requires this now.

The first requirement is to have [`docker`](https://www.docker.com/)
installed locally and running.

Next, change into the build/release directory and run the following command:
```
docker build -t=lnd-release .
```

Once that command has finished executing, our build environment will now be
ready.

At this point, we're now ready to build the latest release. The way the
container is set up, we'll mount the present `lnd` directory _into_ the
container at `/tmp/build/lnd`. This way we can use the existing code checked
in, but ensure the build will be deterministic from the container's
perspective.  After the container build is done, a new file should be
available in the top-level `lnd` directory, which will include the final
release directory.

To build a release, run the following command (from 'build/release`):
```
docker run -it -v $(pwd)/../..:/tmp/build/lnd lnd-release <TAG>
```

Where `<TAG>` is the name of the next release of `lnd`.

> ### Linux

Reviewer: ```suggestion
### Linux/macOS/Windows (WSL)
```

Reviewer: I would also move this section above the docker container since most users will fall under this category.

> No prior set up is needed on Linux if one doesn't wish to use the container

Reviewer: ```suggestion
No prior set up is needed on these systems if one doesn't wish to use the container
```

packaged in this library. Instead, you can build the release binaries from your
`lnd` directory, no matter where it's located. One can download `lnd` to one's
CWD with the following command:
```
git clone https://github.com/lightningnetwork/lnd.git
```

> Afterwards, the release manager/verified simply needs to run: `./release.sh

Reviewer: ```suggestion
Afterwards, the release manager/verifier simply needs to run: `./release.sh
```

# lnd's Reproducible Build System

This package contains the container that the `lnd` project uses in order to

> build binaries for each new release. As go Go 1.13, with some new build

Reviewer: ```suggestion
build binaries for each new release. As of Go 1.13, with some new build
```

> flags, binaries are now reproducible, meaning to developers can build the

Reviewer: ```suggestion
flags, binaries are now reproducible, allowing developers to build the
```

binary on distinct machines, and end up with a byte-for-byte identical

> binary. However, this wasn't _fully_ solved in Go 1.13, as the build system

Reviewer: Do we still need to include this note with the addition of the `-s -w` flags?

> binary. However, this wasn't _fully_ solved in Go 1.13, as the build system

Author: No 1.13 is correct here. 1.13.2 will let us drop the `-buildid=` requirement.

still includes the directory the binary is built into the binary itself,.

> This means that in order for developers/users to produce this binary, it

Author: Don't think so given that we can't get the same binary with Go 1.13 vs 1.13.1.

needs to be built from the _same directory_.

## Building a New Release

### MacOS/Windows

The first requirement is to have [`docker`](https://www.docker.com/)
installed locally and running.

> Next, from the top-level `lnd` directory, run the following command:

Reviewer 3: Needs to say:
> Next, change into the `build/release` directory and run the following command:

Otherwise it's not going to work. The `docker run` command shown later also expects you to be in `build/release`.

```
docker build -t=lnd-release .
```

Once that command has finished executing, our build environment will now be
ready.

At this point, we're now ready to build the latest release. The way the
container is set up, we'll mount the present `lnd` directory _into_ the
container at `/tmp/build/lnd`. This way we can use the existing code checked
in, but ensure the build will be deterministic from the container's
perspective.  After the container build is done, a new file should be
available in the directory the container was run in, which will include the
final release directory.

To build a release, run the following command:
```

> docker run --it -v $(pwd)/../..:/tmp/build/lnd lnd-release <TAG>

Reviewer 3: Needs to be `-it` not `--it`

```

Where `<TAG>` is the name of the next release of `lnd`.

### Linux

No prior set up is needed on Linux if one doesn't wish to use the container

> packaged in this library. Instead, the only requirement is that `lnd` be

Reviewer: Update to remove no longer needed requirement.

cloned into the following directory: `/tmp/build/lnd`.

This can be done with the following command: 
```

> git clone https://github.com/lightningnetwork/lnd.git /tmp/build/

Reviewer: Can remove `/tmp/build` path.

```

> Afterwards, the release manager/verified simply need to run: `./release.sh

Reviewer: ```suggestion
Afterwards, the release manager/verifier simply needs to run: `./release.sh
```

> <TAG>`, where `<TAG>` is the name of this next release/tag.

Reviewer: ```suggestion
<TAG>`, where `<TAG>` is the name of the next release/tag.
```
